Feature/upgrade launcher script

SolomonShorser-OICR opened the pull request:
New scripts:
- for creating new releases.
- for upgrading to new version of architecture-setup.

SolomonShorser-OICR committed:
New work on a script to upgrade a launcher host

SolomonShorser-OICR committed:
cleanup

SolomonShorser-OICR committed:
Update documentation printing.

SolomonShorser-OICR committed:
Do not force repo checkouts.

SolomonShorser-OICR committed:
New file which will create releases.

SolomonShorser-OICR committed:
Updated with new dependencies

SolomonShorser-OICR committed:
Updated with new dependencies

SolomonShorser-OICR committed:
Updated with new dependencies

SolomonShorser-OICR committed:
Updated with new dependencies

SolomonShorser-OICR committed:
Updated with new dependencies

SolomonShorser-OICR committed:
Updated script.

SolomonShorser-OICR committed:
update gitignore

SolomonShorser-OICR committed:
Merge branch 'feature/upgrade_launcher_script' of https://github.com/ICGC-TCGA-PanCancer/architecture-setup into feature/upgrade_launcher_script

briandoconnor committed:
Updated with new dependencies

SolomonShorser-OICR committed:
Revert "Updated with new dependencies"

This reverts commit 6abd8ddbf1597c62e21f56ecdd8fac14e5e36c52.

SolomonShorser-OICR committed:
Update main.yml

SolomonShorser-OICR committed:
Merge branch 'feature/upgrade_launcher_script' of https://github.com/ICGC-TCGA-PanCancer/architecture-setup into feature/upgrade_launcher_script

Conflicts:
	roles/bindle-profiles/vars/main.yml

SolomonShorser-OICR committed:
Updated with new dependencies

SolomonShorser-OICR committed:
Merge branch 'feature/upgrade_launcher_script' of https://github.com/ICGC-TCGA-PanCancer/architecture-setup into feature/upgrade_launcher_script

Conflicts:
	roles/bindle-profiles/vars/main.yml

SolomonShorser-OICR committed:
Some small fixes
Added a "test" mode triggered by flag "--test"
Added a check for the correct version of bash
Fixed usage of base64 to be more platform independent (-w option not available on OS X)

SolomonShorser-OICR committed:
minor cleanup

SolomonShorser-OICR committed:
Add release to architecture-setup
Added code that will create a release tag on architecture-setup, using the same tag name as was used on the other repositories.

SolomonShorser-OICR committed:
Update README.md

SolomonShorser-OICR committed:
Update README.md

Added further details for create_release.sh

SolomonShorser-OICR committed:
Update README.md

Added note about bash 4.x

SolomonShorser-OICR committed:
Added help text, better option handling

SolomonShorser-OICR committed:
Merge branch 'feature/upgrade_launcher_script' of https://github.com/ICGC-TCGA-PanCancer/architecture-setup into feature/upgrade_launcher_script

SolomonShorser-OICR committed:
Update README.md

SolomonShorser-OICR committed:
Small tweaks for OS X

SolomonShorser-OICR committed:
Updated with new dependencies

SolomonShorser-OICR committed:
Update README.md

SolomonShorser-OICR committed:
Update README.md

SolomonShorser-OICR committed:
Code cleanup

denis-yuen commented on README.md:
```diff
@@ -33,3 +33,45 @@ If you wish to setup some other host, you will need to edit your inventory. Repl
+
+    upgrade_architecture_setup.sh -v 1.0.5
+    
+The file `roles/bindle-profiles/vars/main.yml` that is a part of architecture-setup 1.0.5 will be downloaded to the current local repository. Ansible will then be run (`sudo ansible-playbook -i inventory site.yml`).
```
Is it actually necessary to run as sudo here and below?

SolomonShorser-OICR replied:
I think it was necessary when I tried to run it manually, so it would probably be necessary to run as sudo within the script, unless the script itself is run as sudo (currently, I don't do that, but I imagine it would work - I'll have to test that).

denis-yuen replied:
Attach the error message that shows up when you run without sudo, I'm curious what the issue is since individual ansible steps should automatically use sudo as needed.

SolomonShorser-OICR replied:
trying to reproduce now...

SolomonShorser-OICR replied:
Could not reproduce, but I'm not longer working with a fresh & clean machine. It's probably not an issue after all (I will try again later with new instance just to see).

denis-yuen commented on README.md:
```diff
@@ -33,3 +33,45 @@ If you wish to setup some other host, you will need to edit your inventory. Repl
+
+    upgrade_architecture_setup.sh -u MyGithubUserName -v 1.0.5
+
+You will be prompted for the password for the account you used.
```
Is it possible here to use a token instead of a password here? https://help.github.com/articles/creating-an-access-token-for-command-line-use/

Edit to add: it looks like you use a token later for creating new releases, do that here too

SolomonShorser-OICR replied:
My initial understanding of this was that there are other organizations also running the seqware with pancancer workflows on their own fleet of VMs. I don't know if they use github or not, so i didn't want to require them to create a github authentication token since they might not have a github account. This script uses basic authentication to get files from the github API, but it is optional: if someone wants to run this script, then they can do so _without_ authenticating. There _are_ rate limits to unauthenticated requests (something like 60 per hour per IP address), but they seem reasonable if someone is just upgrading a few launchers.

IF we know that everyone who might run this will have a github account, then we can require tokens for authentication.

denis-yuen replied:
Maybe I'm reading the above incorrectly, but I read it as there being two options
1) Run without authentication, doesn't require an account
2) Run with authentication using the "-u" parameter and be prompted for a password. 

It would only be the latter case where I recommend using access tokens instead of a password.

SolomonShorser-OICR replied:
Yes, the current options are: 
1. Run without authentication - doesn't require github accout, API rate limits may apply
2. Run with basic authentication - requires github account, uses "-u" option with script.

I can switch the authentication method to use a token _instead_ of basic auth. - do you think it's still necessary to keep the unauthenticated option?

denis-yuen replied:
Yes, your logic is sound that we need unauthenticated access for cloud shepherds. 
Let's just use tokens instead of basic auth for option 2

SolomonShorser-OICR replied:
I've change the script to do a simple `git checkout <label>` so it shouldn't be necessary to require _any_ kind of authentication anymore. Also, it ensures that the entire local repo is up to date, just in case other files have changed.

denis-yuen commented on roles/bindle-profiles/tasks/main.yml:
```diff
@@ -16,26 +16,31 @@
   git: repo=https://github.com/CloudBindle/Bindle
        dest=/home/{{ user_name }}/architecture2/Bindle
        version={{ bindle_git_branch }}
+       force=no
```
Will this fail silently if there are changes in their repo? Make sure that the script fails loudly if thy have changes

SolomonShorser-OICR replied:
In my tests, this caused ansible to fail and stop when it tried to checkout the repository.

SolomonShorser-OICR replied:
trying to reproduce now...

SolomonShorser-OICR replied:
When a file in any of the repos (in this case, pancancer-bag) is modified, the exact error message is:

<pre>
TASK: [bindle-profiles | Checkout the Bindle git repo] ************************ 
ok: [bindle]

TASK: [bindle-profiles | Checkout the Seqware git repo] *********************** 
ok: [bindle]

TASK: [bindle-profiles | Checkout the pancancer-bag git repo] ***************** 
failed: [bindle] => {"failed": true}
msg: Local modifications exist in repository (force=no).

FATAL: all hosts have already failed -- aborting

PLAY RECAP ******************************************************************** 
           to retry, use: --limit @/home/ubuntu/site.retry

bindle                     : ok=18   changed=5    unreachable=0    failed=1   

ubuntu@ip-10-111-186-42:~/architecture-setup$ 
</pre>


I'm not sure if there's any way to get a clearer message from git (such as the output of `git status`) but this should at least indicate to the user that they need to fix that repo before they can move on.

denis-yuen replied:
Cool

SolomonShorser-OICR replied:
After watching the process take foreeeeeeeeeeeever on one of Adam's launchers, I added some code to do `git status` on the repos before ansible checks them out. This should speed the whole process up.

denis-yuen commented on roles/bindle-profiles/vars/main.yml:
```diff
@@ -1,8 +1,8 @@
 ---
 # The git repo branches, if you're working in production then use the latest stable tagged version
-bindle_git_branch: "2.0-alpha.4"
-seqware_bag_git_branch: "1.0-beta.0"
```
Make sure to revert this to valid versions before closing this pull request. 
For testing purposes, you can override these variables from the CLI http://docs.ansible.com/playbooks_variables.html#passing-variables-on-the-command-line

SolomonShorser-OICR committed:
New script
This script is to be used to update any references to the workflow version

SolomonShorser-OICR committed:
Code cleanup

SolomonShorser-OICR committed:
ignore the github.token file

SolomonShorser-OICR committed:
New script to help update worker nodes
This script will determine which worker nodes are available to have an new version of a worklow pushed to them.

SolomonShorser-OICR committed:
new and modified scripts
New:
  upgrade_worker_nodes.sh - downloads new sanger workflow file, and calls George's playbook to update nodes
  upgrade_launcher_and_workers.sh - wrapper script calls other scripts

SolomonShorser-OICR committed:
minor code cleanup

SolomonShorser-OICR committed:
restored code

code to call ansible for architecture-setup was accidentally deleted - it has now been put back.

SolomonShorser-OICR committed:
Added workflow version

SolomonShorser-OICR committed:
fixed typo

SolomonShorser-OICR committed:
Cleanup for release

adamjohnwright committed:
Update get_nodes_for_update.sh

small bug fix

SolomonShorser-OICR committed:
Add code to check for changed repos before calling ansible

SolomonShorser-OICR committed:
Small bug fix: variable needed to be quoted

SolomonShorser-OICR committed:
Merge branch 'feature/upgrade_launcher_script' of https://github.com/ICGC-TCGA-PanCancer/architecture-setup into feature/upgrade_launcher_script

SolomonShorser-OICR committed:
Update README.md

Added some information about some of the other scripts.

SolomonShorser-OICR committed:
Simplified regex for grep
...also simplified `git status` output to ignore unstaged changes.

SolomonShorser-OICR committed:
Merge branch 'feature/upgrade_launcher_script' of https://github.com/ICGC-TCGA-PanCancer/architecture-setup into feature/upgrade_launcher_script

SolomonShorser-OICR committed:
change: check ALL repos before exiting.

SolomonShorser-OICR committed:
Improved functionality
 - Script now checks files to see if they have diffs before attempting to check int
 - Logging!
 - Prints list of all files actually updated.

SolomonShorser-OICR committed:
Changes and improvements
 - Logging!
 - Changed download roles/bindle-profiles/vars/main.yml to checkout the entire repo (other files might have changed too!)
 - Added check for existence of repo directory

SolomonShorser-OICR committed:
I think there might be a bug in this...
Checking the old method back in so that when I test this with Wei, it shouldn't cause any problems.

SolomonShorser-OICR committed:
Updated with new dependencies

SolomonShorser-OICR committed:
reset to content from main branch.

SolomonShorser-OICR committed:
Changes: automatic release tags, test mode confirmation
Release tags will automatically be created based on the previous release tag, for all repositories. The "release name" argument is no longer needed for this script.
User will be asked to confirm when test mode is NOT set.

SolomonShorser-OICR committed:
Do checkout <label> on architecture-setup
I think it is better to do a full checkout of a specific label of architecture-setup rather than use curl and the github API to download a single updated yml file, because there could be other files that are also updated.
Checking out a label should update the entire local repo, rather than just one file.

SolomonShorser-OICR committed:
Code cleanup.

SolomonShorser-OICR committed:
Added logging functionality.

denis-yuen committed:
Typo

SolomonShorser-OICR committed:
Ignore generated log files.

SolomonShorser-OICR committed:
err codes and check current version
Added code to return explicit error codes
Added a check to check if the current version of architecture-setup is the same as what the user requested.

SolomonShorser-OICR committed:
Exit codes and exit on error
Fixed some scripts to properly return an error exit code.
Changed upgrade_launcher_and_workers.sh to explicitly check (and exit on non-zero) for exits codes from subscripts, instead of using `set -e`

SolomonShorser-OICR committed:
Merge branch 'feature/upgrade_launcher_script' of https://github.com/ICGC-TCGA-PanCancer/architecture-setup into feature/upgrade_launcher_script
